compare_platforms/compare_platforms_read_only.c

```diff
@@ -0,0 +1,147 @@
+    review.realname = realname;
+    review.privatereview = privatereview;
+    review.publicreview = publicreview;
+    review.permissions = permissions;
+    print_details(&review);
+
+#ifdef __CHERI_PURE_CAPABILITY__
+    printf("\nsmallsz=%zx, CRRL(smallsz)=%zx\n", smallsz,
+        __builtin_cheri_round_representable_length(smallsz));
+    printf("biggersz=%zx, CRRL(biggersz)=%zx\n", biggersz,
+        __builtin_cheri_round_representable_length(biggersz));
+    printf("username=%#p realname=%#p diff=%tx\n", username, realname, realname - username);
+    printf("realname=%#p privatereview=%#p diff=%tx\n", realname, privatereview, privatereview - realname);
+    printf("privatereview=%#p publicreview=%#p diff=%tx\n", privatereview, publicreview, publicreview - privatereview);
+    printf("publicreview=%#p permissions=%#p diff=%tx\n", publicreview, permissions, permissions - publicreview);
+    printf("realname=%#p permissions=%#p diff=%tx\n", realname, permissions, permissions - realname);
+#else
+    printf("\nusername=%#p realname=%#p diff=%tx\n", username, realname, realname - username);
+    printf("realname=%#p privatereview=%#p diff=%tx\n", realname, privatereview, privatereview - realname);
+    printf("privatereview=%#p publicreview=%#p diff=%tx\n", privatereview, publicreview, publicreview - privatereview);
+    printf("publicreview=%#p permissions=%#p diff=%tx\n", publicreview, permissions, permissions - publicreview);
+    printf("realname=%#p permissions=%#p diff=%tx\n", realname, permissions, permissions - realname);
+#endif
+
+    assert((size_t)(username + smallsz) <= (size_t)realname); 
+    assert((size_t)(privatereview + biggersz) <= (size_t)publicreview);
+    assert((size_t)(realname + smallsz) <= (size_t)permissions); 
+
+// If we are in a CHERI environment, protect the review by setting
+// the relevant struct members (in this case all of them)
+// to read-only.  
+bool b_improved = false;
+  
+#ifdef __CHERI_PURE_CAPABILITY__
+    if(can_write(&review) == false) {
+        printf("\nSetting the review to read-only.\n");
+        struct review *ro_review = set_read_only(&review);
+        assert((cheri_perms_get(ro_review) & CHERI_PERM_STORE) == 0);
+
+        printf("\nThe struct is read-only so trying to change the review will make the program crash...\n");
+        fflush(stdout);
+        
+        b_improved = try_improve_publicreview(ro_review, biggersz, false);
+        
+        // This line should be unreachable.
+        print_details(ro_review);
+    }
+# else
+    // Try editing the public review in a non-CHERI environment,
+    // by simply checking the write permissions and refusing if none are found.
+    b_improved = try_improve_publicreview(&review, biggersz, false);
+#endif
+
+    // This code will be unreachable in a CHERI environment,
+    // but it would crash whether or not the struct had been set to read-only.
+    if(b_improved == false) {
+        printf("\nOverflowing reviewer realname by 1\n");
```

Comment:
I might be talking gibberish, but I would be careful with overflowing by just 1, due to the "one past the end" rule. I'm unsure if it's applicable here, but just in case.

Reply:
I see your point. In the original tutorial, which was for a riscv64-purecap environment, the overflow succeeded on one past the end. In cheri-morello it doesn't. I think this is interesting enough to keep in. I would like to see if it still behaves differently in riscv64-purecap as well, or if it's the same in both environments.

Reply:
Alright; let's keep this comment here to remind ourselves, and once everything else is in order, we can decide how to proceed.

Reply:
Test

Reply:
Test

Reply:
In [#78 (comment)](https://github.com/capablevms/cheri-examples/pull/78#issuecomment-1404988563) I wrote:

"Asserts refactored to make sense according to the relative positions of the pointers: there should be no overlap, and the overflow-by-1 should come from the pointer on the immediate left of permissions. Fixed in fa66ce07d959169de7f32c137822df458d99bc80"

===

The review struct now has three members of two different sizes: smallsz and biggersz. The bigger-sized pointer is publicreview.

The order of the pointer addresses is different on my local machine (Apple clang), on bencher14 natively (Debian), and on morello-purecap. 

On Apple clang, the biggest pointer has the lowest address. 
On morello-purecap, the biggest pointer has the highest address.
On Debian, the biggest pointer is in the middle.

Previously, the asserts were not taking the different pointer sizes into account. The asserts check that the pointers are (expected to be) non-overlapping according to their allocated sizes. To do this, we have to check the order of the addresses.

The two smaller pointers are realname and permissions. 

In all cases, the permissions are initialized to r. Then the overflow tries to add one or more 'w's.

—

On Apple clang, the order is: the biggest pointer (publicreview) has the lowest address, followed by realname, followed by permissions. 

Because permissions comes after realname, the overflow-by-1 has to come from realname. This is the output:

Overflowing realname into permissions by 1:
Permissions: rw

—

On Debian, realname has the lowest address, followed by the biggest pointer, followed by permissions. 

Because permissions comes after the biggest pointer, the overflow-by-1 therefore has to come from the biggest pointer, publicreview:

Overflowing publicreview into permissions by 1:
Permissions: r

Here, the overflow-by-1 isn't enough to change the permissions.

—

On morello-purecap, realname has the lowest address, followed by permissions, followed by the biggest pointer. 

Here, permissions comes after realname, so the overflow-by-1 comes from realname, just as it did on Apple; but the outcome is a SIGPROT:

Overflowing realname into permissions by 1:
In-address space security exception (core dumped)

Reply:
Alright, that's a lot to unpack. If I understand correctly, this part of the code aims to overwrite a particular read flag to make it writable, by using *some* initial pointer. The issue you observed is that the flag's location is dependent on the struct's in-memory arrangement, which might be different based on local configuration.

So what fa66ce07d959169de7f32c137822df458d99bc80 is attempting to solve is identifying the offset of the flag we want to override from the *some* initial pointer we have?

If all of this is correct, I'd argue this goes against the original intent, which the preamble specifies as "try to exceed the bounds of an object and add write permissions (which will also SIGPROT).". For **this** PR, presuming that the functionality was correct before fa66ce07d959169de7f32c137822df458d99bc80, I would suggest we retract that commit, and merge as is (after CI checks and so on). We can then discuss the refinement you propose, if it's worth adding in a subsequent PR.

Reply:
Okay, that makes sense. I was concerned about leaving too much to chance, but in practice it was working, so I will retract that commit.

Reply:
Fixed in 879f3f56213110b07385348d0802b4a99a53539d